async/logic/Async4BackendRest_DestinationAPI_TopicCheckLogic.java

```diff
@@ -0,0 +1,543 @@
+
+    	return new KafkaProducer(props);
+
+	}
+	
+	private Response ExecDestinationAPIMain(KafkaConsumer<String, String> consumer, KafkaProducer<String, String> producer) {
+
+		StringBuilder destinationStatus = new StringBuilder();
+		StringBuilder destinationResponseBody = new StringBuilder();
+		String receiveMessage = new String();
+		StringBuilder executionTimes = new StringBuilder();
+		StringBuilder sendMessage = new StringBuilder();
+		StringBuilder retryMessage = new StringBuilder();
+		Responses response = new Responses();
+
+		receiveMessage = null;
+		destinationStatus.delete(0, destinationStatus.length());
+		destinationResponseBody.delete(0, destinationResponseBody.length());
+		executionTimes.delete(0, executionTimes.length());
+		sendMessage.delete(0, sendMessage.length());
+		retryMessage.delete(0, retryMessage.length());
+		consumer.subscribe(Arrays.asList(INPUT_TOPIC));
+
+		// メッセージをTopic1から読む
+		ConsumerRecords<String, String> records = consumer.poll(Duration.ofSeconds(60l));
+
+		for (ConsumerRecord<String, String> record : records) {
+			System.out.println(String.format("*record* %s:%s", record.offset(), record.value()));
+			receiveMessage = record.value();
+		}
+
+		// メッセージが拾えなかったら
+		if(receiveMessage == null) {
+			System.out.println("Async4BackendRest_DestinationAPI_Topic relode");
+			//continue;
+			return Response.status(Response.Status.CREATED.getStatusCode()).entity(response).build();
+					}
+
+		Response typeresponse = ExecuteDestinationAPI(receiveMessage, destinationStatus, destinationResponseBody, executionTimes);
+
+		// 代理実行したAPIがエラーの場合
+		if (Integer.parseInt(destinationStatus.toString()) >= Response.Status.MOVED_PERMANENTLY.getStatusCode()) {
+			System.out.println("Destination API Error.");
+			// 実行回数超過の場合
+			if(3 < Integer.parseInt(executionTimes.toString())) {
+				// 超過
+				System.out.println("over 3times");
+
+			} else {
+				// メッセージ再生成(実行回数に1加算する)
+				typeresponse = CreateRetryMessage(receiveMessage, retryMessage, executionTimes);
+					
+				producer.beginTransaction();
+				
+				// Topic１に再登録
+				producer.send(new ProducerRecord<String, String>(INPUT_TOPIC, retryMessage.toString()));
+							
+				Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = new HashMap<>();
+
+				for (TopicPartition partition : records.partitions()) {
+					List<ConsumerRecord<String, String>> partitionedRecords = records.records(partition);
+					long offset = partitionedRecords.get(partitionedRecords.size() - 1).offset();
+
+					offsetsToCommit.put(partition, new OffsetAndMetadata(offset + 1));
+				}
+
+				producer.sendOffsetsToTransaction(offsetsToCommit, CONSUMER_GROUP_ID);
+
+				// トランザクション確定
+				producer.commitTransaction();
+							
+				//continue;
+				return Response.status(Response.Status.CREATED.getStatusCode()).entity(response).build();
+			}
+		}
+		
+		// 送信メッセージ生成
+		typeresponse = CreateSendMessage(receiveMessage, sendMessage, destinationStatus, destinationResponseBody);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *typeresponse*. Execution paths that do not contain closure statements were detected. Ensure *typeresponse* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).